pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,243 @@
+package distributor
+
+import (
+	"context"
+	"fmt"
+	"os"
+	// "net/http"
+	"sync"
+	"time"
+
+	"github.com/weaveworks/common/mtime"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/kvstore"
+)
+
+// Treat these as consts.
+var haInstanceBytes = []byte("prom_ha_instance")
+var haClusterBytes = []byte("prom_ha_cluster")
+
+const prefix = "prom_ha"
+
+// Track the most recent timestamp for the instance we're
+// accepting samples from in a Prometheus HA setup.
+type instanceTracker struct {
+	instance  string
+	timestamp int64
+}
+
+// Track the instance we're accepting samples from
+// for each HA cluster we know about.
+type tracker struct {
+	// Instances we are accepting samples from.
+	electedLock sync.RWMutex
+	elected     map[string]instanceTracker
+
+	// We should only update the timestamp this often
+	writeTimeout int64
+
+	consulLock       sync.RWMutex
+	overwriteTimeout int64 // Timeout that has to occur before we overwrite to a different instance
+	client           kvstore.KVClient
+}
+
+func NewClusterTracker(local bool) clusterTracker {
+	var client kvstore.KVClient
+	var err error
+	codec := kvstore.ProtoCodec{Factory: ProtoInstanceDescFactory}
+	if local {
+		client = kvstore.NewInMemoryKVClient(codec)
+	} else {
+		client, err = kvstore.NewConsulClient(kvstore.ConsulConfig{
+			Host: "localhost:8500",
+		}, codec)
+
+		if err != nil {
+			os.Exit(1)
+		}
+	}
```

Comment:
Can you use `ring.newKVStore` and pass the codec to that too?

Reply:
Also, split out the `ring.Config` into a `KVConfig` that embeds the `Store string` and the `ConsulConfig`, and then put this in your config.  Then have `ring.newKVStore()` take this config.